### ISSUE TITLE
AWS, qemu `cloud_init` klib Issues

### ISSUE THREAD

**Issue by paulreimer** [nanovms/nanos#1774]
I am trying to use the `cloud_init` klib on AWS (my use-case is to avoid having secrets baked in to the image(s) which seems to be the case for `"Env"` variables, and to have flexibility to change variables for each instance and/or reconfigure an existing instance at runtime). So for this, I am thinking to use the AWS EC2 "user data" option, the contents of which are available within the VM via fetching `http://169.254.169.254/latest/user-data`

I am seeing a few issues (which I could split into separate GitHub Issues if preferred):
1. AWS: When using the `cloud_init` klib on AWS, the first two bytes of the `dest` appear to be random binary data (i.e. `0xffff`, or `0x1234`); the rest of the file is correct. I can confirm that fetching the above URL e.g. at run-time using a test `Program` has the fully correct user data. Using a different URL like `http://httpbin.org/uuid` also has the same problem: the first two bytes are corrupt.
2. AWS: The `cloud_init` mechanism only seems to update the `dest` file on the first boot of the instance where the URL returns a 200; i.e. changing the AWS "user data" and rebooting doesn't overwrite the previous file, it remains with the previous data. (perhaps this is by design, though for my purposes I wish it performed the URL fetch on every boot)
3. QEMU: When using the `cloud_init` klib locally / via QEMU, I can't get it to work at all. It seems to hang at boot when using any URL, e.g. `"src": "https://ops.city"`.

Please let me know if there is any other info I can provide or tests that I can run.

**Comment by francescolavra** [comment]
Hi @paulreimer. To address your points:

1. Thanks for catching this issue, it's caused by a bug in the kernel that will be fixed in #1775
2. The cloud_init klib supports a configuration option to overwrite previous files: it's called "overwrite". So if you specify this option for a given file, by inserting an "overwrite" JSON attribute with any string value, cloud_init will re-download the file at every boot. Example:
```
"ManifestPassthrough": {
  "cloud_init": {
    "download": [{"src": "http://169.254.169.254/latest/user-data", "dest": "/user-data", "overwrite": "t"}]
  }
}
```
3. When cloud_init cannot download one or more files, the kernel does not start the user program. The rationale for this is that we want all files to be ready and accessible when the program starts. If you are trying to download a file from a non-existent or unreachable server (e.g. http://169.254.169.254/latest/user-data from within a QEMU guest), the program will never start because cloud_init will try to contact the HTTP server multiple times, with each attempt ending unsuccessfully after a timeout. Also, HTTP chunked transfer encoding is not supported by cloud_init, so if the server uses this encoding, a file download may never complete (and this again makes everything stall); the homepage from https://ops.city is transferred via chunked encoding.

**Comment by paulreimer** [comment]
@francescolavra Thanks for looking into it! Looks like all will be solved with that PR, and if I use `overwrite` for my `download` entry.

FYI I used `https://ops.city` as my test URL as that was the `src` used in the [Klibs `cloud_init` documentation](https://github.com/nanovms/ops-documentation/blob/master/klibs.md?plain=1#L205)
